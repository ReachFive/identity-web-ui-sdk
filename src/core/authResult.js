import { parseJwtTokenPayload } from '../helpers/jwt'
import { logError } from '../helpers/logger'

export function enrichAuthResult(response) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value) {
    return value && (value.accessToken || value.idToken);
}

export function isErrorResponse(value) {
    return value && value.error;
}
