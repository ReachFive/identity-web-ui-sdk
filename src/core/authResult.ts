import { AuthResult, ErrorResponse } from '@reachfive/identity-core';

import { parseJwtTokenPayload } from '../helpers/jwt';
import { logError } from '../helpers/logger';

export function enrichAuthResult(response: AuthResult) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload,
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value: unknown): value is AuthResult {
    return (
        value !== null &&
        typeof value === 'object' &&
        ('accessToken' in value || 'idToken' in value)
    );
}

export function isErrorResponse(value: unknown): value is ErrorResponse {
    return value !== null && typeof value === 'object' && 'error' in value;
}
