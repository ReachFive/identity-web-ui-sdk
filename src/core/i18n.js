import isPlainObject from 'lodash-es/isPlainObject';

export default function resolveI18n(defaultMessages, messages = {}) {
    const mergedMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key, params) => {
        const template = mergedMessages[key] || key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param]), template)
            : template;
    };
}

function flattenObject(object, prefix = []) {
    return isPlainObject(object)
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
