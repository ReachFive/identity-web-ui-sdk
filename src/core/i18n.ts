export type I18nMessages = Record<string, string>

export type I18nNestedMessages = Record<string, string | I18nMessages>

export type I18nMessageParams = Record<string, unknown>

export type I18nResolver = (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => string

export function resolveI18n<T extends (Record<string, T> | string)>(defaultMessages: I18nMessages = {}, messages: I18nNestedMessages = {}): I18nResolver {
    const mergedMessages: I18nMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => {
        const template = mergedMessages[key] ?? fallback?.(params) ?? key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param] as string), template)
            : template;
    };
}

export default resolveI18n

function flattenObject(object: I18nNestedMessages | string, prefix: string[] = []): Record<string, string> {
    return typeof object === "object" && object !== null
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
