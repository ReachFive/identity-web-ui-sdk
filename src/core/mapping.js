import { getValue, setValue } from '../helpers/propertyHelpers';

export class PathMapping {
    constructor(modelPath) {
        this.modelPath = modelPath;
    }

    bind(model) {
        return getValue(model, this.modelPath);
    }

    unbind(model, value) {
        return setValue(model, this.modelPath, value);
    }
}
