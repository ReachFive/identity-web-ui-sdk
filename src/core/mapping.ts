import { ProfileAddress } from '@reachfive/identity-core';
import { getValue, setValue } from '../helpers/propertyHelpers';

export interface PathMapping {
    bind<T extends Record<string, unknown>>(model: T): unknown;
    unbind<T extends Record<string, unknown>, V>(model: T, value: V): T | V;
}

export class DefaultPathMapping implements PathMapping {
    protected readonly modelPath: string;

    constructor(modelPath: string) {
        this.modelPath = modelPath;
    }

    bind<T extends Record<string, unknown>>(model: T) {
        return getValue(model, this.modelPath);
    }

    unbind<T extends Record<string, unknown>, V>(model: T, value: V) {
        return setValue(model, this.modelPath, value);
    }
}

export class AddressPathMapping implements PathMapping {
    protected readonly property: string;

    constructor(property: string) {
        this.property = property;
    }

    private defaultIndex(model: { addresses?: ProfileAddress[] }) {
        const defaultIndex = (model.addresses ?? []).findIndex(address => address.isDefault);
        return defaultIndex > -1 ? defaultIndex : 0;
    }

    bind<T extends { addresses?: ProfileAddress[] }>(model: T) {
        return getValue(model, `addresses.${this.defaultIndex(model)}.${this.property}`);
    }

    unbind<T extends { addresses?: ProfileAddress[] }, V>(model: T, value: V) {
        return setValue(model, `addresses.${this.defaultIndex(model)}.${this.property}`, value);
    }
}
