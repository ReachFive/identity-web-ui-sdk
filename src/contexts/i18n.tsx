import React, { ComponentType, PropsWithChildren, useMemo } from 'react';

import { I18nMessages, I18nNestedMessages, I18nResolver, resolveI18n } from '../core/i18n';

export interface Props {
  defaultMessages?: I18nMessages
  messages?: I18nNestedMessages
}

export const I18nContext = React.createContext<I18nResolver | undefined>(undefined);

export function useI18n(): I18nResolver {
  const context = React.useContext(I18nContext);
  if (!context) {
    throw new Error('No I18nContext provided');
  }

  return context;
}

export interface I18nProps {
  i18n: I18nResolver
}

export type WithI18n<P> = P & I18nProps

export function withI18n<T extends I18nProps = I18nProps>(WrappedComponent: ComponentType<T>) {
  const displayName = WrappedComponent.displayName || WrappedComponent.name || "Component";

  const ComponentWithI18n = (props: Omit<T, keyof I18nProps>) => {
    const i18n = useI18n();
    return <WrappedComponent {...({ i18n })} {...(props as T)} />;
  };

  ComponentWithI18n.displayName = `withI18n(${displayName})`;

  return ComponentWithI18n;
}

export function I18nProvider({ children, defaultMessages, messages }: PropsWithChildren<Props>): JSX.Element | null {
  const resolver = useMemo(() => resolveI18n(defaultMessages, messages), [defaultMessages, messages])
  return <I18nContext.Provider value={resolver}>{children}</I18nContext.Provider>
}
