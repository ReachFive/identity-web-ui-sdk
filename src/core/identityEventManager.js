import { enrichAuthResult } from './authResult'
import EventManager from '../helpers/eventManager'

export function createEventManager() {
    const eventManager = new EventManager();

    return {
        on(eventName, listener) {
            eventManager.on(eventName, listener);
        },

        off(eventName, listener) {
            eventManager.off(eventName, listener)
        },

        fireEvent(eventName, data) {
            if (eventName === 'authenticated') {
                eventManager.fire(eventName, enrichAuthResult(data));
            } else {
                eventManager.fire(eventName, data);
            }
        }
    }
}
