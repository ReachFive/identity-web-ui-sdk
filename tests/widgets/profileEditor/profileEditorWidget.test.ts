/**
 * @jest-environment jest-fixed-jsdom
 */
import { ComponentProps } from 'react';

import { beforeEach, describe, expect, jest, test } from '@jest/globals';
import '@testing-library/jest-dom/jest-globals';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import 'jest-styled-components';

import { type Client, type Profile } from '@reachfive/identity-core';

import { type I18nMessages } from '@/core/i18n';
import { OnError, OnSuccess } from '@/types';
import ProfileEditorWidget from '@/widgets/profileEditor/profileEditorWidget';

import { defaultConfig } from '../renderer';
import { componentGenerator, snapshotGenerator } from '../renderer';

const defaultI18n: I18nMessages = {};

describe('Snapshot', () => {
    const generateSnapshot = (
        options: ComponentProps<typeof ProfileEditorWidget>,
        user: Partial<Profile>
    ) => {
        // @ts-expect-error partial Client
        const apiClient: Client = {
            getUser: jest.fn<Client['getUser']>().mockResolvedValue(user as Profile),
        };

        const generate = snapshotGenerator(ProfileEditorWidget, apiClient, defaultI18n);

        return generate(options);
    };

    describe('profile editor', () => {
        test(
            'basic',
            generateSnapshot(
                {
                    accessToken: 'azerty',
                    fields: ['givenName', 'familyName'],
                },
                {
                    givenName: 'John',
                    familyName: 'Do',
                }
            )
        );
    });
});

describe('DOM testing', () => {
    const getUser = jest.fn<Client['getUser']>();
    const updateProfile = jest.fn<Client['updateProfile']>();

    const onError = jest.fn<OnError>();
    const onSuccess = jest.fn<OnSuccess>();

    beforeEach(() => {
        getUser.mockClear();
        updateProfile.mockClear();
        onError.mockClear();
        onSuccess.mockClear();
    });

    // @ts-expect-error partial Client
    const apiClient: Client = {
        getUser,
        updateProfile,
    };

    const generateComponent = componentGenerator(ProfileEditorWidget, apiClient, defaultI18n);

    describe('profileEditor', () => {
        test('default', async () => {
            const user = userEvent.setup();

            // @ts-expect-error partial Profile
            const profile: Profile = {
                givenName: 'John',
                familyName: 'Do',
                consents: {
                    optinTesting: {
                        granted: false,
                        consentType: 'opt-in',
                        consentVersion: {
                            versionId: 1,
                            language: 'fr',
                        },
                        date: '2021-01-01T00:00:00.000Z',
                    },
                },
            };

            getUser.mockResolvedValue(profile as Profile);

            updateProfile.mockResolvedValue();

            await generateComponent({
                accessToken: 'azerty',
                fields: ['given_name', 'family_name', 'optinTesting'],
                onError,
                onSuccess,
            });

            expect(getUser).toBeCalledWith(
                expect.objectContaining({
                    accessToken: 'azerty',
                    fields: 'givenName,familyName,consents.optinTesting',
                })
            );

            const givenNameInput = screen.getByLabelText('givenName');
            expect(givenNameInput).toBeInTheDocument();
            expect(givenNameInput).toHaveValue(profile.givenName);

            const familyNameInput = screen.getByLabelText('familyName');
            expect(familyNameInput).toBeInTheDocument();
            expect(familyNameInput).toHaveValue(profile.familyName);

            await userEvent.clear(givenNameInput);
            await userEvent.type(givenNameInput, 'alice');
            await userEvent.clear(familyNameInput);
            await userEvent.type(familyNameInput, 'reachfive');

            // const consentCheckbox = screen.getByTestId('consents.optinTesting.1')
            const consentCheckbox = screen.getByLabelText(
                defaultConfig.consentsVersions.optinTesting.versions[0].title
            );
            await user.click(consentCheckbox);

            const submitBtn = screen.getByRole('button', { name: 'save' });
            expect(submitBtn).toBeInTheDocument();

            await user.click(submitBtn);

            expect(updateProfile).toBeCalledWith(
                expect.objectContaining({
                    accessToken: 'azerty',
                    data: {
                        givenName: 'alice',
                        familyName: 'reachfive',
                        consents: {
                            optin_testing: {
                                // consent key should be snakecase
                                granted: true,
                                consentType: 'opt-in',
                                consentVersion: {
                                    versionId: 1,
                                    language: 'fr',
                                },
                            },
                        },
                    },
                })
            );

            expect(onSuccess).toBeCalledWith(expect.objectContaining({ name: 'user_updated' }));
            expect(onError).not.toBeCalled();
        });

        test('api get user failed', async () => {
            getUser.mockRejectedValue(new Error('Unexpected error'));

            await generateComponent(
                {
                    accessToken: 'azerty',
                    fields: ['given_name', 'family_name'],
                    onError,
                    onSuccess,
                },
                {},
                () => expect(getUser).toBeCalled()
            );

            expect(onSuccess).not.toBeCalled();
            expect(onError).toBeCalled();
            expect(onError).toBeCalledWith(new Error('Unexpected error'));
        });

        test('api update user failed', async () => {
            const user = userEvent.setup();

            const profile = {
                givenName: 'John',
                familyName: 'Do',
            };

            getUser.mockResolvedValue(profile as Profile);

            updateProfile.mockRejectedValue('Unexpected error');

            await generateComponent({
                accessToken: 'azerty',
                fields: ['given_name', 'family_name'],
                onError,
                onSuccess,
            });

            expect(getUser).toBeCalled();

            const givenNameInput = screen.getByLabelText('givenName');
            await userEvent.clear(givenNameInput);
            await userEvent.type(givenNameInput, 'alice');

            const familyNameInput = screen.getByLabelText('familyName');
            await userEvent.clear(familyNameInput);
            await userEvent.type(familyNameInput, 'reachfive');

            const submitBtn = screen.getByRole('button', { name: 'save' });
            await user.click(submitBtn);

            expect(updateProfile).toBeCalled();

            expect(onSuccess).not.toBeCalled();
            expect(onError).toBeCalledWith('Unexpected error');
        });
    });
});
