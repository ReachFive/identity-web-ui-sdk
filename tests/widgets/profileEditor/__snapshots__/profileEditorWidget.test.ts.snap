// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot profile editor basic 1`] = `
.c7 {
  color: #adb5bd;
}

.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c1 {
  --color-primary: #229955;
  --color-destructive: #dc4e41;
  --color-background: #ffffff;
  --color-text: #495057;
  --color-border: #ced4da;
  --spacing-padding-y: 9;
  --spacing-padding-x: 12;
  --spacing-block-inner-height: 38;
  --spacing: 10;
  --font-generic: 14;
  --border-width: 1;
  --radius: 3;
}

.c6 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #ffffff;
  background-color: #229955;
  border: 1px solid #229955;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
}

.c6:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(34,153,85,0.5);
}

.c6:hover,
.c6:active {
  background-color: #1b7842;
  border-color: #1b7842;
}

.c6[disabled] {
  opacity: .65;
}

.c2 {
  position: relative;
}

.c8 {
  display: block;
  padding: 9px 12px;
  text-align: center;
}

.c4 {
  color: #495057;
  margin-bottom: 5px;
  display: inline-block;
}

.c3 {
  margin-bottom: 10px;
}

.c3 > label::after {
  content: "\\A0*";
  color: #dc4e41;
}

.c5 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.428571429;
  color: #495057;
  border-radius: 3px;
  border-width: 1px;
  border-style: solid;
  border-color: #ced4da;
  background-color: #fff;
  background-image: none;
  padding: 9px 12px;
  -webkit-transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c5:focus {
  border-color: #5fdb94;
  box-shadow: 0 0 0 3px rgba(95,219,148,0.5);
  outline: 0;
}

.c5:disabled,
.c5[readonly] {
  background-color: #e9ecef;
}

.c5::-webkit-input-placeholder {
  color: #868e96;
}

.c5::-moz-placeholder {
  color: #868e96;
}

.c5:-ms-input-placeholder {
  color: #868e96;
}

.c5::placeholder {
  color: #868e96;
}

<div>
  <div
    class="c0 c1 r5-widget"
    transition="entering"
  >
    <form
      class="c2"
      novalidate=""
    >
      <div
        class="c3"
        required=""
      >
        <label
          class="c4"
          for="givenName"
        >
          givenName
        </label>
        <input
          class="c5"
          data-testid="givenName"
          id="givenName"
          name="givenName"
          placeholder="givenName"
          required=""
          title="givenName"
          value="John"
        />
      </div>
      <div
        class="c3"
        required=""
      >
        <label
          class="c4"
          for="familyName"
        >
          familyName
        </label>
        <input
          class="c5"
          data-testid="familyName"
          id="familyName"
          name="familyName"
          placeholder="familyName"
          required=""
          title="familyName"
          value="Do"
        />
      </div>
      <button
        class="c6"
        data-testid="submit"
        type="submit"
      >
        save
      </button>
      <span
        class="c7 c8"
      >
        *
        form.required.fields
      </span>
    </form>
  </div>
</div>
`;
